### PR TITLE
OSX elimination of IconRef, fully native Bitmap/Mask implementation

### PR DESCRIPTION
**Pull request by csomor**
Removing obsolete IconRef (classic mac) implementation, switching to bitmap implementation which does everything using CGBitmapContext. Right now for testing reasons there's still a switch to turn the 'old' implementation by setting `wxOSX_USE_ICONREF` to 1 and `wxOSX_BITMAP_NATIVE_ACCESS` to 0

### CODE REVIEW

File: include/wx/osx/bitmap.h
Comment (vadz) [inline include/wx/osx/bitmap.h]: I think those should have `OSX` prefix or be declared in `wxMaskBase` (and implemented in the existing `wxMask` implementations deriving from it) to make it more clear that they're not part of the public API or, alternatively, to make them part of the public API.

Otherwise it will be too easy to write code which compiles with wxMac, but not any other ports.

File: include/wx/osx/setup0.h
Comment (vadz) [inline include/wx/osx/setup0.h]: If we intend to keep these options (i.e. if they're not going to be removed before/soon after this PR is merged), we also need to check for them in `include/wx/osx/chkconf.h`.

File: src/osx/carbon/utilscocoa.mm
Comment (vadz) [inline src/osx/carbon/utilscocoa.mm]: It would be be better to avoid checking in commented out code.

File: src/osx/core/dcmemory.cpp
Comment (vadz) [inline src/osx/core/dcmemory.cpp]: Please avoid using `wxT()` in the new code. In all but the most performance-critical places it's ok to omit it entirely, otherwise you can use `wxS()` to avoid an extra conversion, but `wxT()` is never useful.

File: src/osx/core/bitmap.cpp
Comment (vadz) [inline src/osx/core/bitmap.cpp]: It would be nice to avoid overlong (>> 80 characters) lines, they make side-by-side diffs less readable.

File: src/osx/core/bitmap.cpp
Comment (vadz) [inline src/osx/core/bitmap.cpp]: Very minor, but in our style the `{` should be on its own line.

File: src/osx/core/bitmap.cpp
Comment (vadz) [inline src/osx/core/bitmap.cpp]: Minor, but why `1*`? I find this more confusing than helpful...

File: src/osx/core/bitmap.cpp
Comment (vadz) [inline src/osx/core/bitmap.cpp]: Even if it's minimal, why duplicate the code already present above here? I'd write this as

    return const_cast<void*>(const_cast<const wxBitmapRefData*>(this)->GetRawAccess());

**Review by vadz — APPROVED** [review]
Thanks a lot for modernizing this!

All in all I don't have any real comments but it would obviously be even simpler if we could drop the new `wxOSX_USE_XXX` options and just always use the new code. Unless we really need to preserve the old behaviour in some special cases, I'd strongly prefer to do it before even merging this PR.

Thanks again!

**Comment by csomor** [comment]
Hi Vadim, thanks for looking into things, yes, before merge the `wxOSX_BITMAP_NATIVE_ACCESS` will be removed, at least I planned so. I only left it in to allow for easier regression testing in case of bugs. But I think I'll still have to maintain the `wxOSX_USE_ICONREF`, because that's part of the public API, although not documented, or not ?

**Comment by csomor** [comment]
`wxOSX_USE_ICONREF` is right now used in two ways, first as a switch to turn IconRef usage off completely, and then to use IconRef vs NSImage as implementation base (allowing for regression tests). So if no problems arise with the NSImage use, we could still support IconRef externally via the conversion.

**Comment by vadz** [comment]
Sorry, I didn't even realize this. But I blame not just myself (as usual), but also the fact that the code is pretty confusing right now. I think we really need to check for `wxOSX_USE_ICONREF` in 2 places at most (instead of a dozen or more where it is done now): around `wxIconRefIconData` definition and when deciding which kind of `wxIconData` to create. Is there anything preventing us from doing it like this?

BTW, to help with code understanding, I think it would be useful to add `// wxOSX_USE_ICONREF` or `// !wxOSX_USE_ICONREF` comments after the corresponding `#endif`s, at least if they're more than a few lines removed from the matching `#if` or `#else`.

**Comment by csomor** [comment]
Two additional Questions about removing/deprecating:
- I thought about removing SetWidth/Height/Depth for `wxIcon` as well ?
- I've get SetOk on `wxBitmap` and wxIcon and when using native data would not need them at all, no other ports seem to have them. How about direct removal ?

**Comment by csomor** [comment]
OK, sorry for the confusion, I'll do a clean up and differentiate between the two situations in a cleaner way

**Comment by vadz** [comment]
> I thought about removing SetWidth/Height/Depth for `wxIcon` as well ?

If we deprecate them for `wxBitmap`, it makes at least as much sense to do it for `wxIcon` too.

> I've get `SetOk` on wxBitmap and wxIcon and when using native data would not need them at all, no other ports seem to have them. How about direct removal ?

I think it should be fine, nobody seems to be using this one.

File: src/osx/core/bitmap.cpp
Comment (csomor) [reply src/osx/core/bitmap.cpp]: Thanks, done

File: src/osx/core/bitmap.cpp
Comment (csomor) [reply src/osx/core/bitmap.cpp]: True, changed

File: src/osx/core/bitmap.cpp
Comment (csomor) [reply src/osx/core/bitmap.cpp]: Thanks, done

File: src/osx/core/dcmemory.cpp
Comment (csomor) [reply src/osx/core/dcmemory.cpp]: done

File: src/osx/carbon/utilscocoa.mm
Comment (csomor) [reply src/osx/carbon/utilscocoa.mm]: Yes, I'll do so, I'm not sure yet on which level a resize (if at all) should occur

Comment (vadz) [reply src/osx/carbon/utilscocoa.mm]: FWIW, I think that if it's possible not to do it all, it would be preferable.

**Comment by csomor** [comment]
The next question, related to [trac 18146](https://trac.wxwidgets.org/ticket/18146#comment:45) is the position of `wxIcon` vs `wxBitmap`. `wxIcon` is now based on `NSImage`, and since it does not have to provide direct access to the representation bitmap it can keep all appearance support advantages. So it would be the preferred class when using NSImage on OSX IMHO. 
But the current art provider implementation seems to favour bitmaps, even when `GetIcon` is called, it uses `GetBitmap`, which - as long as a bitmap is return from any provider - will be used before the icon bundles will even be considered ...

**Comment by vadz** [comment]
Sorry, what do you mean by "position"? I.e. I'm not sure what is the question here?

And, whatever it is, it would be helpful if you could please explain in a few words (if it's possible) what is the difference between `wxBitmap` and `wxIcon` from macOS point of view? E.g. for MSW, an icon is just a bitmap with a mask (and, just to be complete, a cursor is an icon with a hotspot). But what about macOS, what's the real difference between icons and bitmaps there?

**Comment by csomor** [comment]
OK, `wxBitmap` and `wxIcon` both represent graphics that can be drawn on a `wxDC`. `wxBitmap` offers methods for directly accessing the pixel map data, but `wxIcon` does not - unless converted to a `wxBitmap`. This makes `wxIcon` use an `NSImage` as native representation, which may very late, only at  draw-time decide what representation exactly will be drawn and how. While for a `wxBitmap` I must have / convert into a `CGBitmapContext` and therefore fixate the look in a pixel map, it cannot be adapted to the current appearance etc. later.

**Comment by csomor** [comment]
ok, I should be able to support that, 
this would make wxIcon more or less a subset of wxBitmap on mac, didn't we have that before: ports where wxIcon is kind of a wxBitmap ?

**Comment by vadz** [comment]
I'm late, so I can only say that I agree with Vaclav: ideal would be to change the internal representation only when necessary and use `NSImage` for as long as possible in `wxBitmap` too, as objects of this class are at least as often used as (small) icons as (possibly huge) images.

It's unfortunate that the same class is used for both, of course, but I think by now it's simpler to adapt the implementation than try to change the API...

**Comment by vadz** [comment]
Stefan, would you like me to merge this if it's ready? I'd like to do it after rebasing this on the latest master to get rid of the master merge.

**Comment by vslavik** [comment]
> Hi @vslavik , could you please test whether the current state still works for your app, thanks

I did after your Thursday commits and re-tested now and everything seems fine. 

The only thing that doesn't work is correct rendering of template images in `wxStaticBitmap` — but that didn't work before either and [apparently](https://stackoverflow.com/questions/17148134/nsimage-drawinrect-and-templateimages) can only be fixed by writing a native implementation using `NSImageView`...

**Comment by csomor** [comment]
> I did after your Thursday commits and re-tested now and everything seems fine.
> 
> The only thing that doesn't work is correct rendering of template images in `wxStaticBitmap` — but that didn't work before either and [apparently](https://stackoverflow.com/questions/17148134/nsimage-drawinrect-and-templateimages) can only be fixed by writing a native implementation using `NSImageView`...

@vslavik Perfect, thanks, would you have a sample at hand for the problem ?
then yes @vadz please go ahead (and tell me sometime what I did this time to make things difficult again)

**Comment by vslavik** [comment]
> @vslavik Perfect, thanks, would you have a sample at hand for the problem ?

You mean for template images? They're simply rendered as if they weren't templates at all (i.e. typically black as opposed to gray dynamically chosen for the background). Here's a particularly nasty example, but the same difference exists in light mode:

image

This surprised me too, but template handling is something that is done at the controls level by `NSImageView` or `NSButton`.

**Comment by vadz** [comment]
@csomor Nothing really difficult, but I just wanted to avoid unnecessary merge with master, I think it's preferable to rebase the branch on master instead in such cases.

In any case, I've merged it now, without any changes. I couldn't update `docs/changes.txt` as I don't really know what to put there, i.e. what are the user-visible effects of this change (if any?), so it would be great if you could please do. TIA!

**Comment by csomor** [comment]
> You mean for template images? They're simply rendered as if they weren't templates at all (i.e. typically black as opposed to gray dynamically chosen for the background). Here's a particularly nasty example, but the same difference exists in light mode:
(...)
> This surprised me too, but template handling is something that is done at the controls level by `NSImageView` or `NSButton`.

I've tried using some etched drawing code, and this really looks nice for bigger images, but for small symbols it's not usable. I'll try a native implementation with NSImageView next.

**Comment by vslavik** [comment]
> > Hi @vslavik , could you please test whether the current state still works for your app, thanks
>
> I did after your Thursday commits and re-tested now and everything seems fine.

I think I found a regression after all, in handling of stock icons (previously handled by `LoadIconFromSystemResource`, but I admit I'm a bit confused about which parts in git are used currently and which aren't). Reproducible with e.g. this, which tries to load `wxIcon(wxICON_WARNING, wxBITMAP_TYPE_ICON_RESOURCE)` and fails:
```cxx
wxArtProvider::GetBitmap(wxART_WARNING, wxART_MENU);
```

**Comment by csomor** [comment]
I've checked with version 3.1.1 
```c++
wxArtProvider::GetBitmap(wxART_WARNING, wxART_MENU);
```
does not deliver this icon, due to the fact that first all bitmap providers are looked up, as I mentioned in a post about 3 weeks ago above. But using 
```c++
wxIcon("wxICON_WARNING", wxBITMAP_TYPE_ICON_RESOURCE)
```
gives the native one, and I'll add the code to do so on master, probably with a `LoadFile` extension on the `icns` Handler

**Comment by csomor** [comment]
Hi @vslavik, did the commit improve things ?

**Comment by vslavik** [comment]
@csomor I was *just* testing it — yes, it does (thanks!), although I think it's still subtly different from how it behaved before regarding icon sizes. See e.g. assert window:

screen shot 2018-10-10 at 11 37 40

I'm not actually sure it's an issue with icons loading and not the assert window, though.

**Comment by csomor** [comment]
great, thanks, 
yes, size, that's true, right now I don't resize and get the maximum resolution, resizing would not be a problem, though, the question is when (and where) to do it - or not

**Pull request closed without merging**
